### ISSUE TITLE
Attention and Fuse seem to not improve performance in Shapenet Dataset

### ISSUE THREAD

**Issue by robinp456** [nihalsid/retrieval-fuse#5]
Hi,

I've trained and evaluated the network with the Shapenet data. Somehow the performance decreases in the retrieval and attention phase, is that something that is supposed to happen? The performance increases again in phase 3 (full forward phase). However after training all phases, I evaluated the model's prediction based on attention with retrievals compared to just the unet prediction and the prediction seems to be roughly the same --> The retrievals do not improve the performance. Did you evaluate that aswell on the Shapenet dataset or is there something that might have gone wrong? 

Kind regards

**Comment by nihalsid** [comment]
Hi,

There is a slight performance drop during the retrieval attention phase (phase 2) where we freeze the rest of the networks, but should not be too much. On shapenet with 8^3 -> 64^3 superresolution task the performance of UNet was lower compared to the full optimization in our experiments. For the point cloud -> surface task, the benefits of retrieval are only seen when the point cloud is coarse. If we have a dense point cloud, unet and even SPSR doesnt work too bad. 

Can you share more details of your experiment? Is it the reconstruction task or superresolution. Which config are you trying?

Thanks!

**Comment by robinp456** [comment]
Hey, 
Thanks for the fast reply :)
I used different settings, but for simplicity here are two:

**1. setting:**
experiment: Surface Reconstruction
Data: I used the preprocessed Shapenet  data, provided by the link
Model: I used the pretrained model
config: The surface reconstruction config, provided in the code
I did two evaluations with the metrics.py script:
1. On the validation set (also provided by the link):
 Performance unet: iou:= 0.838, cd:= 0.209, precision:= 0.907, recall:= 0.914, f1:=0.911
 Performance fuse: iou:= 0838, cd:= 0.209, precision:= 0.905, recall:= 0.916, f1:= 0.910
 Performance 1. nearest neighbour: iou:=0,581, cd:=1.104, precision:=0.714, recall:=0.733, 
 f1:=0.724
2. On the test split (created myself (based on your split file) with sdf_gen (target )and sampled 20k points as input)):
 Performance unet: f1:= 0.884
 Performance fuse: f1:= 0.878
Observation: unet and fuse model performance roughly equal

**2. setting**
experiment: Surface Reconstruction
Data: Shapenet Data: Same splits as in your data, but with randomly rotated Shapenet models (generated with sdf gen and sampling points)
config: The surface reconstruction config, provided in the code
model: self trained with the rotated shapenet models
Sadly I dont have all training plots anymore, but here is the plot in phase 2
<img width="802" alt="image" src="https://user-images.githubusercontent.com/38693121/189733741-e6c16173-8326-4c0f-802c-de39de0962ae.png">

--> Observation: Performance drop in Phase 2 (Performance goes up again in Phase 3 (Full Forward))

Best,
Robin

**Comment by robinp456** [comment]
Update: I have two questions regarding the config file: 

1. What exactly is the target and input extend? For what is it needed?
2. Why is the Pointcloud input size 128 while target input size is 64? 
(in base config: input_chunk_size: 128, target_chunk_size: 64)

**Comment by nihalsid** [comment]
> 1. Performance 1. nearest neighbour: iou:=0,581, cd:=1.104, precision:=0.714, recall:=0.733,

One thing that strikes odd is that the retrievals here seem quite bad; is this from the checked in retrievals in the repo? I can see that if the retrieval is so bad, the attention wont help much. 

Regarding questions: 

1. Do you mean attention extent (this is the patch size on which attention operates, see the figure) ? or which key exactly if not this one? 

![screen](https://user-images.githubusercontent.com/932110/189740491-66287bdd-e25e-4be5-b403-0bc61790656c.jpg)

2. The point cloud size is 128 just because we use a voxelized point cloud representation. So, even though the resolution is 128, only very few of these are occupied (points). The target on the other hand represents a shape with proper surface. 

(PS: sorry it's been a while since I saw this code, I have to look at it to remind myself too what's going on)

**Comment by robinp456** [comment]
Thank you very much for your replies! :)

Hmm.. I created the retrievals with the pretrained retrieval model, so the model and the data should be the same as yours. What did you get as performance for the first NN in this setup? Generally speaking, what scores the NN have to achieve in order that the Attention improves the performance?

1. Aaah sorry, I mixed things up: I meant the "patch_context_target: 4,  patch_context_input: 8" in the base config. I think its used in the scene.py file. 
2. I see, but what are the advantage of taking a resolution of 128 instead of 64? Isn't it the same just scaled?

Best,
Robin

**Comment by nihalsid** [comment]
Theoretically should be the same (network or the ones checked in), but just to be 100% sure i would try the one's checked in (maybe its quicker to just evaluate them if its the same?). Typically the retrieval would be only slighly worse than the phase 1 performance to give an improvement. E.g. the retrieval performance on surface reconstruction was really low here, the attention didnt improve things significantly
![screen](https://user-images.githubusercontent.com/932110/189752203-fff8b59a-186c-46c9-88bf-59c91a71f124.jpg)

1. No worries! The patch context there is the input that goes into the retrieval network. E.g. for target where chunks are 16^3 and extent is 4, the actual input to the retrieval is not 16^3 but 24^3, because of neighboring voxel overlap is being used as 'context'. 

2.  64^3 might collapse points to the same voxel if they are two close, losing input information. E.g. lets suppose you have 1D points in the domain [0, 1]. If you use a resolution <=64 if two points are closer than 0.015625, they would merge into one.

**Comment by robinp456** [comment]
That would be nice, 
I see, thank you for the feedback and the explanations, that makes sense :) 
One last follow up question^^ Did you evaluate the influence of the context size in terms of performance?

**Comment by nihalsid** [comment]
Glad it helps! 

I was going through my early logs for this experiment, and I found this one table in the notes on some early scene level experiments. 

![screen](https://user-images.githubusercontent.com/932110/189763450-ea751e0d-53fa-4d79-9d8a-2cc922922d50.jpg)

The receptive field here is the context size; basically from this log and from what I remember the context size = 0 (no receptive field) doesnt work as good as having some non zero context. However, having the global context -> where the entire shape is taken into consideration for the single chunk also works pretty bad. So, basically you want to have some level of neighborhood but not the entire shape as neighborhood, otherwise the network memorizes the position of the chunk relative to the whole shape.
